src/sb16.js

```diff
@@ -0,0 +1,1607 @@
+ * @param{function():number=} handler
+ */
+function register_mixer_read(address, handler)
+{
+    if(!handler)
+    {
+        handler = SB16.prototype.mixer_default_read;
+    }
+    mixer_read_handlers[address] = handler;
+}
+
+/**
+ * @param{number} address
+ * @param{function(number)=} handler
+ */
+function register_mixer_write(address, handler)
+{
+    if(!handler)
+    {
+        handler = SB16.prototype.mixer_default_write;
+    }
+    mixer_write_handlers[address] = handler;
+}
+
+// Reset.
+register_mixer_read(0x00, function()
+{
+    return 0;
+});
+register_mixer_write(0x00);
+
+// Output Stereo Select.
+register_mixer_write(0x0E, function(bits)
+{
+    this.dsp_stereo = bits & 0x2;
+    this.bus.send("speaker-stereo", this.dsp_stereo);
+    this.bus.send("speaker-filter", bits & 0x20);
+});
+
+// IRQ Select.
+register_mixer_read(0x80, function()
+{
+    switch(this.irq)
+    {
+        case 2: return 0x1;
+        case 5: return 0x2;
+        case 7: return 0x4;
+        case 10: return 0x8;
+        default: return 0x0;
+    }
+});
+register_mixer_write(0x80, function(bits)
+{
+    if(bits & 0x1) this.irq = 2;
+    if(bits & 0x2) this.irq = 5;
+    if(bits & 0x4) this.irq = 7;
+    if(bits & 0x8) this.irq = 10;
+});
+
+// DMA Select.
+register_mixer_read(0x81, function()
+{
+    var ret = 0;
+    switch(this.dma_channel_8bit)
+    {
+        case 0: ret |= 0x1; break;
+        case 1: ret |= 0x2; break;
+        case 3: ret |= 0x8; break;
+    }
+    switch(this.dma_channel_16bit)
+    {
+        case 5: ret |= 0x20; break;
+        case 6: ret |= 0x40; break;
+        case 7: ret |= 0x80; break;
+    }
+    return ret;
+});
+register_mixer_write(0x81, function(bits)
+{
+    if(bits & 0x1)  this.dma_channel_8bit  = 0;
+    if(bits & 0x2)  this.dma_channel_8bit  = 1;
+    if(bits & 0x8)  this.dma_channel_8bit  = 3;
+    if(bits & 0x20) this.dma_channel_16bit = 5;
+    if(bits & 0x40) this.dma_channel_16bit = 6;
+    if(bits & 0x80) this.dma_channel_16bit = 7;
+});
+
+// IRQ Status.
+register_mixer_read(0x82, function()
+{
+    var ret = 0x20;
+    for(var i = 0; i < 16; i++)
+    {
+        ret |= i * this.irq_triggered[i];
+    }
+    return ret;
+});
+
+
+
+//
+// FM Handlers
+//
+
+
+
+SB16.prototype.fm_default_write = function(data, register, address)
+{
+    dbg_log("unhandled fm register write. addr:" + register + "|" + h(address) + " data:" + h(data), LOG_SB16);
+    // No need to save into a dummy register as the registers are write-only.
+}
+
+/**
+ * @param{Array} addresses
+ * @param{function(number, number, number)=} handler
+ */
+function register_fm_write(addresses, handler)
+{
+    if(!handler)
+    {
+        handler = SB16.prototype.fm_default_write;
+    }
+    for(var i = 0; i < addresses.length; i++)
+    {
+        fm_handlers[addresses[i]] = handler;
+    }
+}
+
+function between(start, end)
+{
+    if(start === end) return [start];
+    return [start].concat(between(start + 1, end));
```

Comment:
While I like functional programming, this style is massively inefficient in JS and not even more readable, so just use a for loop.